freeipa/src/main/java/com/sequenceiq/freeipa/events/FreeIpaCustomCrnOrNameProvider.java

```diff
@@ -0,0 +1,54 @@
+package com.sequenceiq.freeipa.events;
+
+import static com.sequenceiq.cloudbreak.structuredevent.rest.urlparser.CDPRestUrlParser.RESOURCE_ID;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URLEncodedUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.auth.ThreadBasedUserCrnProvider;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPOperationDetails;
+import com.sequenceiq.cloudbreak.structuredevent.event.rest.RestCallDetails;
+import com.sequenceiq.cloudbreak.structuredevent.rest.CustomCrnOrNameProvider;
+import com.sequenceiq.freeipa.entity.Stack;
+import com.sequenceiq.freeipa.service.stack.StackService;
+
+@Component
+public class FreeIpaCustomCrnOrNameProvider implements CustomCrnOrNameProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaCustomCrnOrNameProvider.class);
+
+    @Inject
+    private StackService stackService;
+
+    @Override
+    public void provide(RestCallDetails restCallDetails, CDPOperationDetails operationDetails, Map<String, String> restParams,
```

Comment:
refactor suggestion: break up into smaller methods
```    @Override
    public void provide(RestCallDetails restCallDetails, CDPOperationDetails operationDetails, Map<String, String> restParams,
            String nameField, String crnField) {
        try {
            Optional<NameValuePair> environmentValue = extractEnvironmentValue(restCallDetails);
            environmentValue.ifPresent(nameValuePair -> updateRestParams(restParams, nameField, crnField, nameValuePair));
        } catch (URISyntaxException e) {
            LOGGER.warn("Cannot provide name and crn because the uri is invalid or any error occurred: {}", e.getMessage(), e);
        }
    }

    private void updateRestParams(Map<String, String> restParams, String nameField, String crnField, NameValuePair nameValuePair) {
        String accountId = ThreadBasedUserCrnProvider.getAccountId();
        Stack stack = stackService.getByEnvironmentCrnAndAccountIdEvenIfTerminated(nameValuePair.getValue(), accountId);
        restParams.put(nameField, stack.getName());
        restParams.put(crnField, stack.getResourceCrn());
        restParams.put(RESOURCE_ID, stack.getId().toString());
    }

    private Optional<NameValuePair> extractEnvironmentValue(RestCallDetails restCallDetails) throws URISyntaxException {
        String requestUri = restCallDetails.getRestRequest().getRequestUri();
        URI uri = new URI(requestUri);
        List<NameValuePair> queryParams = URLEncodedUtils.parse(uri, StandardCharsets.UTF_8);
        Optional<NameValuePair> environmentValue = queryParams.stream().filter(p -> "environment".equals(p.getName())).findFirst();
        return environmentValue;
    }
```